### PR TITLE
Introduce ComponentTagHelper

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/13726

### CODE REVIEW

File: src/Components/Samples/BlazorServerApp/Pages/_Host.cshtml [hunk 1]
```diff
@@ -13,9 +13,7 @@
     <link href="css/site.css" rel="stylesheet" />
 </head>
 <body>
-    <app>
-        @(await Html.RenderComponentAsync<App>(RenderMode.ServerPrerendered))
-    </app>
+    <component type="typeof(App)" render-mode="ServerPrerendered" />
```
Comment (rynowak) [inline src/Components/Samples/BlazorServerApp/Pages/_Host.cshtml]: Should we make the tag helper enforce use as a self-closing tag?

The reason why I'm asking is that someone might do:
```razor
<component ... >
  some stuff
</component>
```

And have some kind of expectations about what that means - currently it means nothing, and we can make that explicitly an error.

File: src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            await base.InitializeAsync();
+
+            // Capture the last log timestamp so that we can filter logs when we
+            // check for duplicate connections.
+            var lastLog = Browser.Manage().Logs.GetLog(LogType.Browser).LastOrDefault();
+            if (lastLog != null)
+            {
+                LastLogTimeStamp = lastLog.Timestamp;
+            }
```
Comment (rynowak) [inline src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs]: Unused though?

File: src/Components/test/testassets/TestServer/Pages/ComponentWithParameters.cshtml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+@page
+
+<component type="typeof(ComponentWithParameters)"
+           render-mode="ServerPrerendered"
+           parameter-Param1="ComponentWithParameters.TestModelValues"
+           parameter-Param2="ComponentWithParameters.DerivedModelValue"
```
Comment (rynowak) [inline src/Components/test/testassets/TestServer/Pages/ComponentWithParameters.cshtml]: This is kinda an interesting case. It's interesting to think about because the same is not true with `Param1`. If you put an instance of `DerivedModel` in that list, it wouldn't roundtrip.

I think is is probably the right thing to do but I'm sure it's going to come up. This use case for serialization is really different from our usage elsewhere. We'll have to make sure to write good docs for this - the main thing is that it's not really possible for people to marshal some kinds of data (like polymorphic collections) without doing a lot of S.T.J extensibility.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.AspNetCore.Mvc.ViewFeatures.RazorComponents;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// A <see cref="TagHelper"/> that renders a Blazor component.
+    /// </summary>
+    [HtmlTargetElement("component", Attributes = ComponentTypeName)]
```
Comment (rynowak) [inline src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: should we enforce the style here?

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            var componentRenderer = ViewContext.HttpContext.RequestServices.GetRequiredService<IComponentRenderer>();
+            var result = await componentRenderer.RenderComponentAsync(ViewContext, ComponentType, RenderMode, Parameters);
```
Comment (rynowak) [inline src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: Do you want to try and prevent the dictionary allocation here?

File: src/Mvc/Mvc.ViewFeatures/src/HtmlHelperComponentExtensions.cs [hunk 6]
```diff
@@ -1,127 +0,0 @@
...
-using Microsoft.AspNetCore.Mvc.ViewFeatures;
-using Microsoft.AspNetCore.Mvc.ViewFeatures.RazorComponents;
-using Microsoft.Extensions.DependencyInjection;
-
-namespace Microsoft.AspNetCore.Mvc.Rendering
-{
-    /// <summary>
-    /// Extensions for rendering components.
-    /// </summary>
-    public static class HtmlHelperComponentExtensions
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/HtmlHelperComponentExtensions.cs]: um.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.RazorComponents
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs]: do we need this new namespace? `.ViewFeatures` is already kinda a junk namespace.

File: src/Mvc/Mvc.ViewFeatures/src/HtmlHelperComponentExtensions.cs (hunk 6 above)
Comment (rynowak) [reply src/Mvc/Mvc.ViewFeatures/src/HtmlHelperComponentExtensions.cs]: just a folder move, ok.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.RazorComponents
+{
+    internal interface IComponentRenderer
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs]: What do we get out of this? If we just called the HTML helper from the tag helper, what are the problmes?

**Review by ryanbrandenburg — COMMENTED** [review]
Neat.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs (hunk 4 above)
Comment (DamianEdwards) [reply src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: "Razor component" vs. "Blazor component"?
Technically, as it is right now, they're Razor Components.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// A <see cref="TagHelper"/> that renders a Blazor component.
+    /// </summary>
+    [HtmlTargetElement("component", Attributes = ComponentTypeName)]
+    public class ComponentTagHelper : TagHelper
+    {
+        private const string ComponentParameterName = "parameters";
+        private const string ComponentParameterPrefix = "parameter-";
```
Comment (DamianEdwards) [inline src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: Perhaps we should shorten this to just `params` and `param-`? Seems well enough known in this case that it would be worthwhile given the character reduction.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs (hunk 8 above)
Comment (DamianEdwards) [reply src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs]: Assuming we at least add the non-generic overload to the HTML helper's existing method right?

Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs]: Having the things as extensions makes it really difficult to test. Also the extension method is just a place to dump code. Outside of using the `ViewContext`, none of the `IHtmlHelper` is used.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs (hunk 7 above)
Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs]: Yeah, we could clean it up. I thought everything else in that directory was intentionally under the namespace, but it's just this type: https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.ViewFeatures/src/RazorComponents/StaticComponentRenderer.cs#L18

File: src/Components/test/testassets/TestServer/Pages/ComponentWithParameters.cshtml (hunk 3 above)
Comment (pranavkm) [reply src/Components/test/testassets/TestServer/Pages/ComponentWithParameters.cshtml]: Yup. I wanted to at least capture some of the ordinary cases. The experience is generally going to be bad with polymorphism. `Dictionary<string, object>` for instance are going to round trip as `JsonElement` instances. I'll make sure to specifically say that the implementation uses S.T.J to marshal the values so we recommend simple values that are easy to round trip, don't exceed the size limits, etc

File: src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs (hunk 2 above)
Comment (pranavkm) [reply src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs]: Yup. Copied it from the other test, will remove

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs (hunk 9 above)
Comment (pranavkm) [reply src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: It does sound concise, but we also spent preview8 removing abbreviations from the API surface, so 🤷‍♂

File: src/Mvc/samples/MvcSandbox/Pages/Components.cshtml [hunk 10]
```diff
@@ -15,8 +14,8 @@
     <link href="css/site.css" rel="stylesheet" />
 </head>
 <body>
-    <app>@(await Html.RenderComponentAsync<MvcSandbox.Components.App>(RenderMode.Static))</app>
+    <component type="typeof(MvcSandbox.Components.App)"render-mode="Static" />
```
Comment (SteveSandersonMS) [inline src/Mvc/samples/MvcSandbox/Pages/Components.cshtml]:
```suggestion
    <component type="typeof(MvcSandbox.Components.App)" render-mode="Static" />
```

File: src/Mvc/samples/MvcSandbox/Pages/Components.cshtml [hunk 11]
```diff
@@ -15,8 +14,8 @@
...
 
-    <script src="_framework/components.server.js"></script>
+    <script src="~/_framework/blazor.server.js"></script>
```
Comment (SteveSandersonMS) [inline src/Mvc/samples/MvcSandbox/Pages/Components.cshtml]: Why add the `~/`? As long as the base href is set correctly, it will work without it. And if the base href isn't set correctly, other things will be broken.

File: src/Mvc/samples/MvcSandbox/Views/Shared/_Layout.cshtml [hunk 12]
```diff
@@ -19,6 +19,9 @@
                         <li class="nav-item">
                             <a class="nav-link text-dark" asp-area="" asp-page="/PagesHome">Pages Home</a>
                         </li>
+                        <li class="nav-item">
+                        <a class="nav-link text-dark" asp-area="" asp-page="/Components">Components</a>
```
Comment (SteveSandersonMS) [inline src/Mvc/samples/MvcSandbox/Views/Shared/_Layout.cshtml]:
```suggestion
                            <a class="nav-link text-dark" asp-area="" asp-page="/Components">Components</a>
```

File: src/Mvc/samples/MvcSandbox/Pages/Components.cshtml (hunk 11 above)
Comment (pranavkm) [reply src/Mvc/samples/MvcSandbox/Pages/Components.cshtml]: I was messing with something and this was accidental. I'll revert this 👍

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs (hunk 9 above)
Comment (pranavkm) [reply src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: Updated

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            if (htmlHelper is null)
+            {
+                throw new ArgumentNullException(nameof(htmlHelper));
+            }
+
+            var viewContext = htmlHelper.ViewContext;
+            var componentRenderer = viewContext.HttpContext.RequestServices.GetRequiredService<IComponentRenderer>();
+            return componentRenderer.RenderComponentAsync(viewContext, typeof(TComponent), renderMode, parameters);
+        }
+    }
```
Comment (rynowak) [inline src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs]: We should also add non-generic overloads. The HTML helper is a little too limiting.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    /// <summary>
+    /// A <see cref="TagHelper"/> that renders a Razor component.
+    /// </summary>
+    [HtmlTargetElement("component", Attributes = ComponentTypeName, TagStructure = TagStructure.WithoutEndTag)]
+    public class ComponentTagHelper : TagHelper
```
Comment (rynowak) [inline src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: Should we seal? Would this do something reasonable if someone inherited? I don't have a strong feeling either way.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            var componentRenderer = ViewContext.HttpContext.RequestServices.GetRequiredService<IComponentRenderer>();
```
Comment (rynowak) [inline src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: Should this be a constructor parameter?

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs (hunk 13 above)
Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs]: I thought about it, but in general we wouldn't encourage users to use the html helper if there's an equally capable tag helper available. Do you feel strongly about adding the additional overload?

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs (hunk 14 above)
Comment (pranavkm) [reply src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: We could do it. We've generally left all our tag helpers unsealed in the past, but that's because most of our types are unsealed by default.

File: src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs (hunk 15 above)
Comment (pranavkm) [reply src/Mvc/Mvc.TagHelpers/src/ComponentTagHelper.cs]: I didn't want to make the utility type public as yet, at least until we got enough feedback about how users want to use it and how it fits for the client scenarios that were punted until 5.1. This forces the interface to not be a ctor parameter.

File: src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        public ComponentWithParametersTest(
+            BrowserFixture browserFixture,
+            BasicTestAppServerSiteFixture<PrerenderedStartup> serverFixture,
+            ITestOutputHelper output)
+            : base(browserFixture, serverFixture, output)
+        {
+        }
+
+        [Fact]
+        public void PassingParametersToComponentsWorks()
```
Comment (javiercn) [inline src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs]: There's already an E2E test for this.

File: src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            var parameter1 = Browser.FindElement(By.CssSelector(".Param1"));
+            Assert.Equal(100, parameter1.FindElements(By.CssSelector("li")).Count);
+            Assert.Equal("99 99", parameter1.FindElement(By.CssSelector("li:last-child")).Text);
+
+            // The assigned value is of a more derived type than the declared model type. This check
+            // verifies we use the actual model type during round tripping.
+            var parameter2 = Browser.FindElement(By.CssSelector(".Param2"));
+            Assert.Equal("Value Derived-Value", parameter2.Text);
+
+            // This check verifies CaptureUnmatchedValues works
```
Comment (javiercn) [inline src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs]: Why are we validating this here, it feels completely unrelated to the feature or to the ability to pass parameters to the root component?

File: src/Components/test/testassets/TestServer/Pages/MultipleComponents.cshtml [hunk 18]
```diff
@@ -10,26 +10,26 @@
...
-                @(await Html.RenderComponentAsync<GreeterComponent>(RenderMode.ServerPrerendered))
+                <component type="typeof(GreeterComponent)" render-mode="Server"/>
+                <component type="typeof(GreeterComponent)" render-mode="ServerPrerendered"/>
                 <p>Some content after</p>
             </div>
         </div>
         <div id="container">
-            @(await Html.RenderComponentAsync<GreeterComponent>(RenderMode.Server, new { Name = "Albert" }))
-            @(await Html.RenderComponentAsync<GreeterComponent>(RenderMode.ServerPrerendered, new { Name = "Abraham" }))
+            <component type="typeof(GreeterComponent)" render-mode="Server" param-name='"Albert"' />
```
Comment (javiercn) [inline src/Components/test/testassets/TestServer/Pages/MultipleComponents.cshtml]: Leave some HtmlHelper instances around so that we can validate they keep working.

File: src/Components/test/testassets/TestServer/Pages/PrerenderedHost.cshtml [hunk 19]
```diff
@@ -1,13 +1,14 @@
 @page
 @using BasicTestApp.RouterTest
+
```
Comment (javiercn) [inline src/Components/test/testassets/TestServer/Pages/PrerenderedHost.cshtml]: nit: unnecessary white-space.

File: src/Mvc/Mvc.ViewFeatures/src/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs [hunk 20]
```diff
@@ -206,8 +205,9 @@ internal static void AddViewServices(IServiceCollection services)
             services.TryAddSingleton<SaveTempDataFilter>();
 
             //
-            // Component prerendering
+            // Component rendering
             //
+            services.TryAddScoped<IComponentRenderer, ComponentRenderer>();
```
Comment (javiercn) [inline src/Mvc/Mvc.ViewFeatures/src/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs]: Can we get rid of the interface an use an internal abstract class with a public virtual method instead?

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs (hunk 8 above)
Comment (javiercn) [reply src/Mvc/Mvc.ViewFeatures/src/RazorComponents/IComponentRenderer.cs]: Can we at least get rid of the interface and use an abstract class instead?

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Rendering;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class ComponentRenderer : IComponentRenderer
```
Comment (javiercn) [inline src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs]: Are there any changes to what was there in the HtmlHelper?

File: src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs (hunk 13 above)
Comment (javiercn) [reply src/Mvc/Mvc.ViewFeatures/src/Rendering/HtmlHelperComponentExtensions.cs]: I don't really care that much about the TagHelper, but it should be uber cheap to add this, so I'm with @rynowak on this. There's not even need to add or change tests for them, simply chain the implementations. I trust you know what to do :)

**Review by javiercn — APPROVED** [review]
Left a couple of comments, but otherwise looks solid. The main 3 things are:
* Get rid of IComponentRenderer.
* Add the non-generic HtmlHelper overloads.
* Leave around some HtmlHelper calls in the E2E tests to make sure they keep working E2E.

File: src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs (hunk 21 above)
Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs]: Nope, this was pretty much copied from the old code as-is.

File: src/Mvc/Mvc.ViewFeatures/src/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs (hunk 20 above)
Comment (pranavkm) [reply src/Mvc/Mvc.ViewFeatures/src/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs]: Is there a particular benefit to changing this? Given it's internal, we should be free to modify the contract in the future if we need to.

File: src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs (hunk 17 above)
Comment (pranavkm) [reply src/Components/test/E2ETest/ServerExecutionTests/ComponentWithParametersTest.cs]: Steve noted in the other PR - https://github.com/aspnet/AspNetCore/pull/14465#discussion_r329495458, that we want to ensure we have the ability to pass CaptureUnmatchedValues around. I wanted to make sure we covered that in the context of using the tag helper.
